### PR TITLE
Petsc complex tests

### PR DESCRIPTION
**Pull request by davydden**
that's WIP as `PETSC_WITH_COMPLEX` is not defined in unit tests. I am figuring out how to introduce an extra `DEAL_II_WITH_PETSC_COMPLEX`.

In reference to https://github.com/dealii/dealii/issues/2033

### CODE REVIEW

**Comment by davydden** [comment]
@tamiko could you please point me to the right direction in cmake configure system?

File: include/deal.II/grid/tria.h [hunk 1]
```diff
@@ -39,7 +39,7 @@ DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
 #include <vector>
 #include <list>
 #include <map>
-#include <algorithm>
+#include <numeric>
```
Comment (bangerth) [inline include/deal.II/grid/tria.h]: I don't think you need this anymore. Just rebase to master and this should go away.

Comment (davydden) [inline include/deal.II/grid/tria.h]: sure, did not notice this

**Comment by davydden** [comment]
btw, the complex PETSc wont' compile at the moment, as still need to extend things related to MatrixTools which i overlooked last time. So, please, don't merge this yet.

**Comment by bangerth** [comment]
OK. I'll mark the patch as "WIP". Just remove this from its title (and let us know) once you think it's ready to merge.

**Comment by davydden** [comment]
This PR is relevant again. Still don't know how to introduce a new CMake variable which will be available after `DEAL_II_INITIALIZE_CACHED_VARIABLES()` and thus could be used in unit tests.

**Comment by davydden** [comment]
@tamiko ping

**Comment by davydden** [comment]
i tried to do `#cmakedefine DEAL_II_PETSC_WITH_COMPLEX` in config.h.in and set `SET(DEAL_II_PETSC_WITH_COMPLEX ${PETSC_WITH_COMPLEX})` in configure_3_petsc.cmake, but this variable is not available in unit tests.

**Comment by tamiko** [comment]
> i tried to do #cmakedefine DEAL_II_PETSC_WITH_COMPLEX in config.h.in and set SET(DEAL_II_PETSC_WITH_COMPLEX ${PETSC_WITH_COMPLEX}) in configure_3_petsc.cmake, but this variable is not available in unit tests.

This approach should work. ```config.h``` is sourced in our testsuite. Or do you want to selectively disable tests? (I haven't thought about a way to do this yet).

**Comment by tamiko** [comment]
(Sorry for the late reply. It just took me 10 minutes to find this :-D - I have way too many github notifications per day at the moment.)

**Comment by davydden** [comment]
> This approach should work. config.h is sourced in our testsuite. 

i need it not for C++, but for CMake. I try to disable complete category with

```
IF(DEAL_II_WITH_PETSC AND NOT DEAL_II_PETSC_WITH_COMPLEX)
    DEAL_II_PICKUP_TESTS()
  ENDIF()
```

I think `config.h` is not being read in `INITIALISE_CACHED_VARIABLE()` which is called in each `CMakeList.txt` for a category.

> Or do you want to selectively disable tests? (I haven't thought about a way to do this yet).

It's fine just to enable/disable complete category for now.

**Comment by davydden** [comment]
@tamiko ping :smile:

**Comment by tamiko** [comment]
@davydden The testsuite is implemented as a project - this means that cmake has only the information provided in ```cmake/config/Config.cmake.in``` available.

```cmake/config/CMakeLists.txt``` only exports ```DEAL_II_WITH_<FEATURE>```

**Comment by davydden** [comment]
how can i export additionally something like `DEAL_II_WITH_PETSC_COMPLEX`?

**Comment by tamiko** [comment]
@davydden PR incoming.

**Comment by davydden** [comment]
@tamiko Thanks a lot !!!

File: include/deal.II/base/config.h.in [hunk 2]
```diff
@@ -145,6 +145,9 @@
 /* cmake/modules/FindARPACK.cmake */
 #cmakedefine DEAL_II_ARPACK_WITH_PARPACK
 
+/* cmake/modules/FindPETSc.cmake */
+#cmakedefine DEAL_II_PETSC_WITH_COMPLEX
+
```
Comment (tamiko) [inline include/deal.II/base/config.h.in]: Remove the configuration in ```config.h.in``` please - we won't need it.

File: cmake/configure/configure_3_petsc.cmake [hunk 3]
```diff
@@ -138,3 +138,4 @@ ENDMACRO()
 
 
 CONFIGURE_FEATURE(PETSC)
+SET(DEAL_II_PETSC_WITH_COMPLEX ${PETSC_WITH_COMPLEX})
```
Comment (tamiko) [inline cmake/configure/configure_3_petsc.cmake]: keep this.

File: tests/petsc/CMakeLists.txt [hunk 4]
```diff
@@ -1,6 +1,6 @@
 CMAKE_MINIMUM_REQUIRED(VERSION 2.8.9)
 INCLUDE(../setup_testsubproject.cmake)
 PROJECT(testsuite CXX)
-IF(DEAL_II_WITH_PETSC)
+IF(DEAL_II_WITH_PETSC AND NOT DEAL_II_PETSC_WITH_COMPLEX)
```
Comment (tamiko) [inline tests/petsc/CMakeLists.txt]: keep this

**Comment by davydden** [comment]
in some tests i have 
```
9: --------------------------------------------------------
9: An error occurred in line <255> of file </Users/davydden/libs-sources/deal.ii/davydden/source/lac/petsc_matrix_base.cc> in function
9:     void dealii::PETScWrappers::MatrixBase::compress(const VectorOperation::values)
9: The violated condition was:
9:     last_action == VectorOperation::unknown || last_action == operation
9: The name and call sequence of the exception was:
9:     ExcMessage("Missing compress() or calling with wrong VectorOperation argument.")
9: Additional Information:
9: Missing compress() or calling with wrong VectorOperation argument.
9: --------------------------------------------------------
```
every time `deal.II/lac/petsc_sparse_matrix.h` is included (even when not used). Hmm... @oneliefleft any ideas?

**Comment by davydden** [comment]
never-mind, i am fixing it now...

**Comment by davydden** [comment]
currently
```
The following tests FAILED:
	 44 - petsc_complex/fe_get_function_values.mpirun=3.release (Failed)
	 45 - petsc_complex/parallel_sparse_matrix_01.mpirun=3.debug (Failed)
	 46 - petsc_complex/parallel_sparse_matrix_01.mpirun=3.release (Failed)
	 47 - petsc_complex/solver_real_01.debug (Failed)
	 48 - petsc_complex/solver_real_01.release (Failed)
	 49 - petsc_complex/solver_real_02.debug (Failed)
	 50 - petsc_complex/solver_real_02.release (Failed)
```

**Comment by davydden** [comment]
`parallel_sparse_matrix_01` has to do with the output of `is_symmetric` and `is_hermitian`

**Comment by davydden** [comment]
what's left failing is `petsc_complex/fe_get_function_values.mpirun=3.release`. There could be some issues in Trilinos as Valgrind complains:

```
==58897== Conditional jump or move depends on uninitialised value(s)
==58897==    at 0x10B345F93: dealii::IndexSet::make_trilinos_map(ompi_communicator_t* const&, bool) const (index_set.cc:520)
==58897==    by 0x10B059D85: dealii::TrilinosWrappers::MPI::Vector::reinit(dealii::IndexSet const&, ompi_communicator_t* const&, bool) (trilinos_vector.cc:228)
==58897==    by 0x10B05CDC0: dealii::TrilinosWrappers::MPI::Vector::reinit(dealii::IndexSet const&, dealii::IndexSet const&, ompi_communicator_t* const&, bool) (trilinos_vector.cc:374)
==58897==    by 0x10002DE2F: test() (fe_get_function_values.cc:98)
==58897==    by 0x10002F122: main (fe_get_function_values.cc:174)
```

**Comment by davydden** [comment]
Any hints from Trilinos specialists @kronbichler @Rombur  if this could be the reason for SEGV i see?

**Comment by kronbichler** [comment]
It's strange that only the release test fails for get_function_values. The valgrind message definitely points at a problem because there should be no such problem. Could you please run valgrind on the debug test to see if the same message (or maybe something a bit more verbose) comes up? In particular, I would be interesting in learning what the uninitialized value should be. The only comparison we make around line 520 is the check 'n_elements()>0' which should never be problematic for a well-constructed element. Or is the problem the bool 'overlapping' - but then again we always set a well-defined value ('true') as the function argument.

**Comment by davydden** [comment]
@kronbichler thanks for having a look. That's already valgrind output for debug version. I checked it in debug to see the line numbers. 

In release, i have more complaints from valgrind, but there is not the same for conditional jump. There are some warnings like
```
==58854== 32 bytes in 1 blocks are possibly lost in loss record 5,185 of 11,607
==58854==    at 0x1000545CB: malloc (in /usr/local/Cellar/valgrind/HEAD/lib/valgrind/vgpreload_memcheck-amd64-darwin.so)
==58854==    by 0x11A3CC7DD: operator new(unsigned long) (in /usr/lib/libc++.1.dylib)
==58854==    by 0x10F1C82C1: Epetra_MpiComm::Clone() const (in /usr/local/Cellar/trilinos/12.4.2/lib/libepetra.12.4.2.dylib)
==58854==    by 0x10F11A910: Epetra_BlockMapData::Epetra_BlockMapData(long long, int, long long, Epetra_Comm const&, bool) (in /usr/local/Cellar/trilinos/12.4.2/lib/libepetra.12.4.2.dylib)
==58854==    by 0x10F116CB9: void Epetra_BlockMap::ConstructUserConstant<int>(int, int, int const*, int, int, Epetra_Comm const&, bool) (in /usr/local/Cellar/trilinos/12.4.2/lib/libepetra.12.4.2.dylib)
==58854==    by 0x10F1134D8: Epetra_BlockMap::Epetra_BlockMap(int, int, int const*, int, int, Epetra_Comm const&) (in /usr/local/Cellar/trilinos/12.4.2/lib/libepetra.12.4.2.dylib)
==58854==    by 0x10F186B64: Epetra_Map::Epetra_Map(int, int, int const*, int, Epetra_Comm const&) (in /usr/local/Cellar/trilinos/12.4.2/lib/libepetra.12.4.2.dylib)
==58854==    by 0x1071308F5: dealii::IndexSet::make_trilinos_map(ompi_communicator_t* const&, bool) const (in /Users/davydden/libs-sources/deal.ii/builds/complex/lib/libdeal_II.8.5.0-pre.dylib)
==58854==    by 0x10703C5A8: dealii::TrilinosWrappers::MPI::Vector::reinit(dealii::IndexSet const&, ompi_communicator_t* const&, bool) (in /Users/davydden/libs-sources/deal.ii/builds/complex/lib/libdeal_II.8.5.0-pre.dylib)
==58854==    by 0x10002B8B5: test() (in /Users/davydden/libs-sources/deal.ii/builds/complex/tests/petsc_complex/fe_get_function_values.release/fe_get_function_values.release)
==58854==    by 0x10002CEDD: main (in /Users/davydden/libs-sources/deal.ii/builds/complex/tests/petsc_complex/fe_get_function_values.release/fe_get_function_values.release)
```

**Comment by davydden** [comment]
there is also this, that points to Trilinos 

```
==58854== 113 bytes in 1 blocks are definitely lost in loss record 9,884 of 11,607
==58854==    at 0x1000545CB: malloc (in /usr/local/Cellar/valgrind/HEAD/lib/valgrind/vgpreload_memcheck-amd64-darwin.so)
==58854==    by 0x11B5F3735: opal_os_path (in /usr/local/Cellar/open-mpi/1.10.1_1/lib/libopen-pal.13.dylib)
==58854==    by 0x11B57A153: orte_ess_base_app_abort (in /usr/local/Cellar/open-mpi/1.10.1_1/lib/libopen-rte.12.dylib)
==58854==    by 0x11B578D96: orte_errmgr_base_abort (in /usr/local/Cellar/open-mpi/1.10.1_1/lib/libopen-rte.12.dylib)
==58854==    by 0x122389BD6: abort_peers (in /usr/local/Cellar/open-mpi/1.10.1_1/lib/openmpi/mca_errmgr_default_app.so)
==58854==    by 0x1190F73DA: ompi_mpi_abort (in /usr/local/Cellar/open-mpi/1.10.1_1/lib/libmpi.12.dylib)
==58854==    by 0x1193E776E: PetscSignalHandlerDefault (in /usr/local/Cellar/petsc/3.6.3/complex/lib/libpetsc.3.6.3.dylib)
==58854==    by 0x1193E7AA9: PetscSignalHandler_Private (in /usr/local/Cellar/petsc/3.6.3/complex/lib/libpetsc.3.6.3.dylib)
==58854==    by 0x23805BC0A: ???
==58854==    by 0x10703C5B7: dealii::TrilinosWrappers::MPI::Vector::reinit(dealii::IndexSet const&, ompi_communicator_t* const&, bool) (in /Users/davydden/libs-sources/deal.ii/builds/complex/lib/libdeal_II.8.5.0-pre.dylib)
==58854==    by 0x10002B8B5: test() (in /Users/davydden/libs-sources/deal.ii/builds/complex/tests/petsc_complex/fe_get_function_values.release/fe_get_function_values.release)
==58854==    by 0x10002CEDD: main (in /Users/davydden/libs-sources/deal.ii/builds/complex/tests/petsc_complex/fe_get_function_values.release/fe_get_function_values.release)
```

**Comment by bangerth** [comment]
I would try to get a back trace in the debugger for the release case. This should at least help show where exactly the problem is happening and what statement it is.

**Comment by davydden** [comment]
points to `TrilinosWrappers::MPI::Vector::reinit`:
```
(lldb) bt
* thread #1: tid = 0xc14304, 0x000000010283c373 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(Epetra_Map const&, bool) + 19, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=2, address=0x7fff5f000050)
  * frame #0: 0x000000010283c373 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(Epetra_Map const&, bool) + 19
    frame #1: 0x000000010283c5b8 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(dealii::IndexSet const&, ompi_communicator_t* const&, bool) + 88
    frame #2: 0x000000010002b8b6 fe_get_function_values.release`test() + 502
    frame #3: 0x000000010002cede fe_get_function_values.release`main + 334
    frame #4: 0x00007fff92f615ad libdyld.dylib`start + 1
    frame #5: 0x00007fff92f615ad libdyld.dylib`start + 1
```

**Comment by bangerth** [comment]
Is there a way to get the line in which this happens, and maybe the value of the variables that appear?

**Comment by davydden** [comment]
i don't quite know how to print the line numbers in **release** mode or examine variables. 

Looking at the code of `reinit`
```
 void
    Vector::reinit (const IndexSet &parallel_partitioner,
                    const MPI_Comm &communicator,
                    const bool      omit_zeroing_entries)
    {
      nonlocal_vector.reset();

      Epetra_Map map = parallel_partitioner.make_trilinos_map (communicator,
                                                               true);
      reinit (map, omit_zeroing_entries);
    }
```
and at the backtrace, it seems that it breaks right after creating `Epetra_Map`.

**Comment by davydden** [comment]
i would say something is wrong with Epetra_map
```
(lldb) f 1
frame #1: 0x000000010283c5b8 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(dealii::IndexSet const&, ompi_communicator_t* const&, bool) + 88
libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit:
    0x10283c5b8 <+88>:  leaq   -0x40(%rbp), %rdi
    0x10283c5bc <+92>:  callq  0x10323dd6c               ; libdeal_II.8.5.0-pre.dylib.__TEXT.__stubs + 165774
    0x10283c5c1 <+97>:  addq   $0x18, %rsp
    0x10283c5c5 <+101>: popq   %rbx
(lldb) p map
(__NSAtom *) $0 = 0x0000007200000001
```
My guess is that `frame 1` sits on this line `reinit (map, omit_zeroing_entries);`.

**Comment by bangerth** [comment]
Hm, good question. Can you try to recompile release mode with `-g` (in addition to all of the other optimization flags, of course)? This should allow you more information from inside the debugger...

**Comment by davydden** [comment]
@bangerth good idea! will do...

**Comment by davydden** [comment]
here is more info on the segmentation fault. 
```
(x86_64)
Process 62927 stopped
* thread #1: tid = 0xc28321, 0x0000000102833a93 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(Epetra_Map const&, bool) [inlined] std::__1::enable_if<(is_move_constructible<std::__1::__shared_weak_count*>::value) && (is_move_assignable<std::__1::__shared_weak_count*>::value), void>::type std::__1::swap<std::__1::__shared_weak_count*>(std::__1::__shared_weak_count*&, std::__1::__shared_weak_count*&) at type_traits:3562, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=2, address=0x7fff5f000050)
    frame #0: 0x0000000102833a93 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(Epetra_Map const&, bool) [inlined] std::__1::enable_if<(is_move_constructible<std::__1::__shared_weak_count*>::value) && (is_move_assignable<std::__1::__shared_weak_count*>::value), void>::type std::__1::swap<std::__1::__shared_weak_count*>(std::__1::__shared_weak_count*&, std::__1::__shared_weak_count*&) at type_traits:3562
   3559                                     is_nothrow_move_assignable<_Tp>::value)
   3560 {
   3561     _Tp __t(_VSTD::move(__x));
-> 3562     __x = _VSTD::move(__y);
   3563     __y = _VSTD::move(__t);
   3564 }
   3565
(lldb) bt
* thread #1: tid = 0xc28321, 0x0000000102833a93 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(Epetra_Map const&, bool) [inlined] std::__1::enable_if<(is_move_constructible<std::__1::__shared_weak_count*>::value) && (is_move_assignable<std::__1::__shared_weak_count*>::value), void>::type std::__1::swap<std::__1::__shared_weak_count*>(std::__1::__shared_weak_count*&, std::__1::__shared_weak_count*&) at type_traits:3562, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=2, address=0x7fff5f000050)
  * frame #0: 0x0000000102833a93 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(Epetra_Map const&, bool) [inlined] std::__1::enable_if<(is_move_constructible<std::__1::__shared_weak_count*>::value) && (is_move_assignable<std::__1::__shared_weak_count*>::value), void>::type std::__1::swap<std::__1::__shared_weak_count*>(std::__1::__shared_weak_count*&, std::__1::__shared_weak_count*&) at type_traits:3562
    frame #1: 0x0000000102833a93 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(Epetra_Map const&, bool) [inlined] std::__1::shared_ptr<Epetra_MultiVector>::swap(std::__1::shared_ptr<Epetra_MultiVector>&) at memory:4611
    frame #2: 0x0000000102833a93 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(Epetra_Map const&, bool) [inlined] std::__1::shared_ptr<Epetra_MultiVector>::reset() at memory:4619
    frame #3: 0x0000000102833a93 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(this=0x00007fff5f000000, input_map=0x00007fff5fbfdbf0, omit_zeroing_entries=false) + 19 at trilinos_vector.cc:204
    frame #4: 0x0000000102833cd8 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(this=<unavailable>, parallel_partitioner=<unavailable>, communicator=<unavailable>, omit_zeroing_entries=<unavailable>) + 88 at trilinos_vector.cc:230
    frame #5: 0x000000010002b8b6 fe_get_function_values.release`test() + 502 at fe_get_function_values.cc:93
    frame #6: 0x000000010002cede fe_get_function_values.release`main(argc=<unavailable>, argv=<unavailable>) + 334 at fe_get_function_values.cc:174
    frame #7: 0x00007fff92f615ad libdyld.dylib`start + 1
    frame #8: 0x00007fff92f615ad libdyld.dylib`start + 1
(lldb) 
```
i'll see if I can find the explanation...

**Comment by davydden** [comment]
Does not make sense, the error is essentially in resetting a shared pointer to `Epetra_MultiVector`
```
(lldb) f 3
frame #3: 0x0000000102833a93 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(this=0x00007fff5f000000, input_map=0x00007fff5fbfdbf0, omit_zeroing_entries=false) + 19 at trilinos_vector.cc:204
   201      Vector::reinit (const Epetra_Map &input_map,
   202                      const bool        omit_zeroing_entries)
   203      {
-> 204        nonlocal_vector.reset();
   205 
   206        if (vector->Map().SameAs(input_map)==false)
   207          vector.reset (new Epetra_FEVector(input_map));
```

**Comment by davydden** [comment]
here's a minimum Trilinos test with a single reinit call (no complex PETSc) that fails for me with `mpirun=3` in release mode for deal.II build with **both complex and real valued PETSc** 
```
#include "../tests.h"

#include <deal.II/base/logstream.h>
#include <deal.II/base/index_set.h>
#include <deal.II/grid/grid_generator.h>
#include <deal.II/dofs/dof_handler.h>
#include <deal.II/dofs/dof_tools.h>
#include <deal.II/fe/fe_q.h>

#include <deal.II/lac/trilinos_vector.h>

#include <deal.II/distributed/tria.h>
#include <deal.II/distributed/grid_refinement.h>

using namespace dealii;

static const unsigned int dim = 2;

void test ()
{
  MPI_Comm mpi_communicator (MPI_COMM_WORLD);

  parallel::distributed::Triangulation<dim> tria(mpi_communicator,
                                                 typename Triangulation<dim>::MeshSmoothing
                                                 (Triangulation<dim>::smoothing_on_refinement |
                                                  Triangulation<dim>::smoothing_on_coarsening));

  GridGenerator::hyper_cube (tria, -1,0);
  tria.refine_global (2);

  const unsigned int poly_degree = 1;
  FE_Q<dim> fe(poly_degree);

  DoFHandler<dim> dof_handler(tria);
  dof_handler.distribute_dofs(fe);

  IndexSet locally_owned_dofs = dof_handler.locally_owned_dofs ();

  TrilinosWrappers::MPI::Vector vector_Re;
  vector_Re.reinit(locally_owned_dofs, mpi_communicator);

  deallog << "OK" << std::endl;
}


int main (int argc, char *argv[])
{
  std::ofstream logfile("output");
  deallog.attach(logfile);
  deallog.depth_console(0);
  deallog.threshold_double(1.e-10);

  Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);
  {
    test ();
  }
}
```
anyone has any ideas?

**Comment by davydden** [comment]
This is ready for re-review. I replaed the offending part in the unit test with `parallel::distributed::Vector` (that makes sense anyway as we want to test complex PETSc irrespectively of availability of Trilinos). 

In order to keep track of the (possible) problem I created a new Trilinos unit test for `reinit()`, which still fails for me locally but perhaps it will be ok on testers. In either case, this strange issue with `reinit()` seems to be absolutely separate from complex-valued PETSc, so I suggest to merge this PR and dealii with this problem in a separate issue/PR.

**Comment by davydden** [comment]
@Rombur if I am not mistaken, you recently modified `IndexSet::make_trilinos_map()`, the Valgrind issue above in `make_trilinos_map()` https://github.com/dealii/dealii/pull/2230#issuecomment-193207483 could be related. Also the segmentation fault appears right after `make_trilinos_map()`, here's frame 4:
```
frame #4: 0x000000010285e288 libdeal_II.8.5.0-pre.dylib`dealii::TrilinosWrappers::MPI::Vector::reinit(this=<unavailable>, parallel_partitioner=<unavailable>, communicator=<unavailable>, omit_zeroing_entries=<unavailable>) + 88 at trilinos_vector.cc:230
   227 
   228        Epetra_Map map = parallel_partitioner.make_trilinos_map (communicator,
   229                                                                 true);
-> 230        reinit (map, omit_zeroing_entries);
   231      }
   232 
   233 
```
Frame 3 is https://github.com/dealii/dealii/pull/2230#issuecomment-193255510 .
Could you please check if the unit test https://github.com/davydden/dealii/commit/302817dd40526cd8859c3e86b9ecf6969549bfb7 works for you with master?

**Comment by Rombur** [comment]
I will check but now that I think of it, after I made that change Nicola Giuliani reported that he needed to compile the release mode with -g for the tests to pass (see https://groups.google.com/forum/#!searchin/dealii/trilinos|sort:date/dealii/SxVTHGxQ92s/O9cKWW18GgAJ) Can you try adding -g in release mode.

**Comment by davydden** [comment]
@Rombur yes, it still fails with `-g` in Release mode only. Without `-g` i wouldn't be able to have any info on code lines.

**Comment by davydden** [comment]
in other words, it was failing both with `-g` and without it.

**Comment by Rombur** [comment]
It works for me on master

**Comment by davydden** [comment]
that's strange... @drwells @luca-heltai could any of you guys please check if https://github.com/davydden/dealii/commit/302817dd40526cd8859c3e86b9ecf6969549bfb7 passes in release mode with OS-X on master?

**Comment by drwells** [comment]
I do not have access to any OS X machines right now. Do you just want to see if it works with clang?

**Comment by davydden** [comment]
@drwells i must have misinterpreted your comment on another thread. But in any case, it could be related to clang vs gcc.

**Comment by davydden** [comment]
current 8.4 branch passes the test, so it must be introduced in last couple of weeks. I will start bisecting...

**Comment by Rombur** [comment]
@davydden I took your branch and your test passes on my machine. Your test is also very similar to ```mpi/trilinos_vector_reinit``` and that test passes on the tester.

**Comment by davydden** [comment]
@Rombur by bisection i figured out that the bug is introduced in https://github.com/dealii/dealii/commit/a1229b381c03e3d5401234debe96543193581110. I think it is related to `MPI_LOGICAL`, which is a Fortran data type, AFAIK. In C++ there is `MPI::BOOL`. I don't think the MPI standard specifies a bool type for C case, that you use. I would say one should either use `MPI_INT` or C++ bindings with `MPI::BOOL`.

**Comment by davydden** [comment]
p.s. the test fails on ef0bc63fcd885f5c6d015d004f86dfd6208da7f1 and works on
7538693d125a92d180c5e57f286b644fbfbbb1c7 . There is only one commit inbetween :smile: .

**Comment by bangerth** [comment]
Ah, how fascinating! Good detective work. This may then simply depend on pure chance and the MPI implementation you use. I would say change the code so that the variable is an `int` with value either zero or one, and do a plus or max operation as appropriate.

This kind of entirely pernicious bug is so subtle that I'd love to have your testcase checked in along with any fix. Even if none of us can reproduce this issue on our systems, it may be reproducible on someone else's system in the future if we ever accidentally change this back.

Out of curiosity, are there other places in the library where we use `MPI_LOGICAL`?

**Comment by Rombur** [comment]
@bangerth no it seems to be the only place.

**Comment by davydden** [comment]
i will make a PR with a fix to INT and the unit test now...

**Comment by davydden** [comment]
This PR is again ready for re-review.

**Comment by davydden** [comment]
```
100% tests passed, 0 tests failed out of 72
```

**Comment by drwells** [comment]
@davydden I don't think that you need it at this point but I feel inclined to follow up: I can compile, but not run deal.II with clang *and Trilinos* on my machines without getting odd linker errors, so I cannot really help right now (though I am trying to figure out my own problems).

**Comment by davydden** [comment]
@drwells thanks for the follow up! I don't know if that will be of any help, but you can see how `Trilinos` is compiled in `Homebrew` https://github.com/Homebrew/homebrew-science/blob/master/trilinos.rb . And here are instructions used by Luca https://gitlab.com/configurations/cluster-config .

File: tests/petsc_complex/00.cc [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  const PetscScalar alpha = 1.0 + 2.0*PETSC_i;
+
+  // make beta
+  const PetscScalar beta  = 2.0 + 1.0*PETSC_i;
+
+  // multiply them together
+  const PetscScalar gamma = alpha * beta;
+
+  // Check real access
+  Assert (PetscRealPart (gamma)==0., ExcInternalError());
```
Comment (Rombur) [inline tests/petsc_complex/00.cc]: Assert only works in Debug mode. Your test checks nothing in Release.

File: tests/petsc_complex/00.cc [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+  Assert (PetscImaginaryPart (gamma) ==5., ExcInternalError());
+
```
Comment (Rombur) [inline tests/petsc_complex/00.cc]: Same here and below.

File: tests/petsc_complex/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,6 @@
+CMAKE_MINIMUM_REQUIRED(VERSION 2.8.9)
```
Comment (Rombur) [inline tests/petsc_complex/CMakeLists.txt]: We normally require 2.8.8. Is there a reason why you use 2.8.9?

File: tests/petsc_complex/vector_02.cc [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  for (unsigned int k=0; k<v.size(); ++k)
+    v(k) = k;
+
+  v.compress (VectorOperation::insert);
+
+  PETScWrappers::Vector v2(s);
+  for (unsigned int k=0; k<v2.size(); ++k)
+    {
+      // This fails
+      // v2(k) = PetscScalar (k,-k);
```
Comment (Rombur) [inline tests/petsc_complex/vector_02.cc]: Does this fail because ```k``` is an ```unsigned int```?

File: tests/slepc_complex/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,6 @@
+CMAKE_MINIMUM_REQUIRED(VERSION 2.8.9)
```
Comment (Rombur) [inline tests/slepc_complex/CMakeLists.txt]: Why cmake 2.8.9?

**Comment by Rombur** [comment]
Please replace ```Assert``` by ```AssertThrow``` in your tests.

File: tests/petsc_complex/vector_02.cc (hunk 8 above)
Comment (davydden) [inline tests/petsc_complex/vector_02.cc]: that i don't know but it does fail indeed if you try to use this line.

Comment (davydden) [inline tests/petsc_complex/vector_02.cc]: you were right, it had to do with `unsigned int`. I suppose `-k` led to something like max unsigned int.

**Comment by davydden** [comment]
done.

**Pull request merged**